### PR TITLE
[WIP]: ENH/TST: Integrate pytest plugin for enhanced doctesting

### PR DESCRIPTION
**Pull request by Sheila-nk**
*Work-in-progress, DO NOT MERGE*
#### Reference issue

This PR is part of my ongoing internship at Quansight Labs, where I was tasked with building a reusable pytest plugin that provides enhanced doctesting for PyData libraries, specifically Scipy and Numpy. 
It builds on the work done by @ev-br to split the refguide-check doctesting utility into a separate package: https://github.com/scipy/scipy/pull/16391.

#### What does this implement/fix?

This PR introduces a pytest plugin that facilitates the execution of doctests within Scipy's public modules through `dev.py`.
To execute doctests, you can run the following commands on the Scipy shell:
`python dev.py test --doctests` to run all doctests
`python dev.py test --doctests -s {submodule}` for running individual doctests within specific submodules.

#### Additional information

*Pending Work*
- [x] filter out all deprecation warnings
- [ ] handle failures due to output formatting
- [x] handle syntax errors in numpy ufuncs
- ...

### CODE REVIEW

File: .github/workflows/linux_meson.yml
Comment (ev-br) [inline .github/workflows/linux_meson.yml:126]: why a specific hash?

File: dev.py
Comment (ev-br) [inline dev.py:699]: As long as you have a special `doctest` command, this should not be needed.

File: dev.py
Comment (ev-br) [inline dev.py:784]: Also this

File: dev.py
Comment (ev-br) [inline dev.py:1212]: I wonder if there's room for deduplicating `Test` and `Doctest` tasks.

**Review by ev-br — COMMENTED** [review]
Initial style review.

File: .github/workflows/linux_meson.yml
Comment (Sheila-nk) [reply .github/workflows/linux_meson.yml:126]: I wanted to run doctests with the most recent branch of scpdt, especially the one that deduplicates test items and exempts the tests directory.

File: dev.py
Comment (Sheila-nk) [reply dev.py:699]: I wasn't sure about completely removing the option of running doctests using the `python dev.py test --doctests` incantation. I think now we have settled on: `python dev.py doctest`?

File: dev.py
Comment (Sheila-nk) [reply dev.py:1212]: Let me look into it.

File: .github/workflows/linux_meson.yml
Comment (ev-br) [reply .github/workflows/linux_meson.yml:128]: OK to keep for the time being; we'll need to transfer the whole package off my anyway. Will need to change it here when https://github.com/ev-br/scpdt/pull/91/ lands.

File: dev.py
Comment (ev-br) [reply dev.py]: I personally on a fence on which of the two way makes it. Not the two though.

**Comment by ev-br** [comment]
Looks like a great start @Sheila-nk !

Here's what I think next steps should be:

- [ ] make sure doctesting run on CI
- [ ] make sure the plugin collects a superset of `refguide-check` log; IOW that we are not missing objects
- [ ] there are lots of doctest failures locally. This PR is currently missing filters and other customization from gh-16391, these may help somewhat. Other failures need investigating.
- [ ] at least locally, I still need the env var, `$ PY_IGNORE_IMPORTMISMATCH=1  python dev.py doctest`
.
The external dependency on scpdt: 
- [ ] needs to get a better name (doctest_sp, doctest_scipy, scipy_doctest, something else?)
- [ ] needs to get out of my GH
- [ ] we need to decide on the way to use it: git submodule or optional test dependency
- [ ] the noise in `$ python dev.py shell && pytest --pyargs scipy` needs filtering out

File: dev.py
Comment (rgommers) [reply dev.py:1210]: Have you considered making this `python dev.py test --doctests` rather than a separate command? That way you may be able to just set the `pytest_args` you need, and not change anything else.

File: scipy/_lib/_testutils.py
Comment (rgommers) [inline scipy/_lib/_testutils.py:80]: @ev-br can't that `_interpnd_info.py` file simply be deleted? It isn't used at all and hasn't seen any nontrivial change since it was written 13 years ago says `git lg --follow scipy/interpolate/_interpnd_info.py`.

File: .github/workflows/linux_meson.yml
Comment (rgommers) [inline .github/workflows/linux_meson.yml:124]: It'd be preferable for this to replace the refguide check job here I think: https://github.com/scipy/scipy/blob/95cc0fd65011187c2589eaeddf0f0c86d04cc318/.circleci/config.yml#L170-L176. We don't need both.

Comment (ev-br) [reply .github/workflows/linux_meson.yml:124]: Maybe best move the refguide-check to GHA and combine with this one. Note that not all of refguide-check is obsoleted: the part which checks consistency between `__all__` lists and refguide is still needed (or if not, it's a separate discussion anyway).

Comment (rgommers) [reply .github/workflows/linux_meson.yml:124]: Ah, I didn't realize that we still need it for that, thanks.

> Maybe best move the refguide-check to GHA and combine with this one.

I think the point is that for docs-only PRs you can run with `[skip actions]` and trigger only the docs-related checks in CircleCI. If it's here, there's no good way to avoid running ~20 other jobs.

File: dev.py
Comment (Sheila-nk) [reply dev.py:1210]: This works. It does solve the duplication problem.

File: scipy/_lib/_testutils.py
Comment (ev-br) [reply scipy/_lib/_testutils.py]: tl;dr I'd rather keep it.

I think this file implements (a part) of computations of coefficients used in https://github.com/scipy/scipy/blob/main/scipy/interpolate/interpnd.pyx#L392
So it's a form of documentation, an addition to the long comment at https://github.com/scipy/scipy/blob/main/scipy/interpolate/interpnd.pyx#L443, similar to what we have in scipy.special, https://github.com/scipy/scipy/tree/main/scipy/special/_precompute

File: dev.py
Comment (ev-br) [reply dev.py:785]: We'll need to not forget to document in big bold letters that without the plugin `pytest --doctest-modules` runs both tests and doctests, while with the plugin, it's only doctests.

File: scipy/_lib/_testutils.py
Comment (ev-br) [inline scipy/_lib/_testutils.py:80]: Minor: `_lib` ignore should be a tad more selective.

Comment (Sheila-nk) [reply scipy/_lib/_testutils.py:80]: It's because the `_lib` modules were being collected for doctesting on running: `python dev.py test --doctests -v`. I'm not sure if we want it included.
```
scipy/_lib/_disjoint_set.py::scipy._lib._disjoint_set.DisjointSet PASSED                                                                                                                             [  0%]
scipy/_lib/_docscrape.py::scipy._lib._docscrape.namedtuple FAILED                                                                                                                                    [  0%]
scipy/_lib/_finite_differences.py::scipy._lib._finite_differences.arange FAILED                                                                                                                      [  0%]
scipy/_lib/_finite_differences.py::scipy._lib._finite_differences.array FAILED                                                                                                                       [  0%]
scipy/_lib/_finite_differences.py::scipy._lib._finite_differences.hstack FAILED                                                                                                                      [  0%]
scipy/_lib/_finite_differences.py::scipy._lib._finite_differences.prod FAILED                                                                                                                        [  0%]
scipy/_lib/_gcutils.py::scipy._lib._gcutils.gc_state FAILED                                                                                                                                          [  0%]
scipy/_lib/_gcutils.py::scipy._lib._gcutils.assert_deallocated FAILED                                                                                                                                [  0%]
scipy/_lib/_tmpdirs.py::scipy._lib._tmpdirs.in_dir FAILED                                                                                                                                            [  0%]
scipy/_lib/_tmpdirs.py::scipy._lib._tmpdirs.in_tempdir FAILED                                                                                                                                        [  0%]
scipy/_lib/_tmpdirs.py::scipy._lib._tmpdirs.tempdir FAILED                                                                                                                                           [  0%]
scipy/_lib/_util.py::scipy._lib._util.AxisError FAILED                                                                                                                                               [  0%]
scipy/_lib/_util.py::scipy._lib._util.Generator PASSED                                                                                                                                               [  0%]
scipy/_lib/_util.py::scipy._lib._util.Generator.binomial FAILED                                                                                                                                      [  0%]
scipy/_lib/_util.py::scipy._lib._util.Generator.bytes FAILED                                                                                                                                         [  0%]
scipy/_lib/_util.py::scipy._lib._util.Generator.chisquare FAILED                                                                                                                                     [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.choice FAILED                                                                                                                                        [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.dirichlet FAILED                                                                                                                                     [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.exponential FAILED                                                                                                                                   [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.f FAILED                                                                                                                                             [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.gamma FAILED                                                                                                                                         [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.geometric FAILED                                                                                                                                     [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.gumbel FAILED                                                                                                                                        [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.hypergeometric FAILED                                                                                                                                [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.integers FAILED                                                                                                                                      [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.laplace FAILED                                                                                                                                       [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.logistic FAILED                                                                                                                                      [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.lognormal FAILED                                                                                                                                     [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.logseries FAILED                                                                                                                                     [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.multinomial FAILED                                                                                                                                   [  1%]
scipy/_lib/_util.py::scipy._lib._util.Generator.multivariate_hypergeometric FAILED                                                                                                                   [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.multivariate_normal FAILED                                                                                                                           [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.negative_binomial FAILED                                                                                                                             [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.noncentral_chisquare FAILED                                                                                                                          [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.noncentral_f FAILED                                                                                                                                  [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.normal FAILED                                                                                                                                        [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.pareto FAILED                                                                                                                                        [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.permutation FAILED                                                                                                                                   [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.permuted FAILED                                                                                                                                      [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.poisson PASSED                                                                                                                                       [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.power FAILED                                                                                                                                         [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.random FAILED                                                                                                                                        [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.rayleigh FAILED                                                                                                                                      [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.shuffle FAILED                                                                                                                                       [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.spawn FAILED                                                                                                                                         [  2%]
scipy/_lib/_util.py::scipy._lib._util.Generator.standard_cauchy FAILED                                                                                                                               [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.standard_exponential FAILED                                                                                                                          [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.standard_gamma FAILED                                                                                                                                [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.standard_normal FAILED                                                                                                                               [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.standard_t FAILED                                                                                                                                    [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.triangular FAILED                                                                                                                                    [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.uniform FAILED                                                                                                                                       [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.vonmises FAILED                                                                                                                                      [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.wald FAILED                                                                                                                                          [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.weibull FAILED                                                                                                                                       [  3%]
scipy/_lib/_util.py::scipy._lib._util.Generator.zipf FAILED                                                                                                                                          [  3%]
scipy/_lib/decorator.py::scipy._lib.decorator.n_args PASSED                                                                                                                                          [  3%]
scipy/_lib/doccer.py::scipy._lib.doccer.docformat FAILED                                                                                                                                             [  3%]
scipy/_lib/doccer.py::scipy._lib.doccer.inherit_docstring_from FAILED                                                                                                                                [  3%]
scipy/_lib/doccer.py::scipy._lib.doccer.indentcount_lines FAILED                                                                                                                                     [  3%]
scipy/_lib/doccer.py::scipy._lib.doccer.unindent_string FAILED                                                                                                                                       [  4%]
scipy/_lib/uarray.py::scipy._lib.uarray.Dispatchable FAILED                                                                                                                                          [  4%]
scipy/_lib/uarray.py::scipy._lib.uarray.all_of_type FAILED                                                                                                                                           [  4%]
scipy/_lib/uarray.py::scipy._lib.uarray.determine_backend FAILED                                                                                                                                     [  4%]
scipy/_lib/uarray.py::scipy._lib.uarray.determine_backend_multi FAILED                                                                                                                               [  4%]
scipy/_lib/uarray.py::scipy._lib.uarray.generate_multimethod FAILED                                                                                                                                  [  4%]
scipy/_lib/uarray.py::scipy._lib.uarray.mark_as FAILED                                                                                                                                               [  4%]
scipy/_lib/_uarray/__init__.py::scipy._lib._uarray FAILED 
```

File: dev.py
Comment (Sheila-nk) [reply dev.py:785]: Yes, I have opened an issue on updating the documentation: https://github.com/ev-br/scpdt/issues/96

**Comment by ev-br** [comment]
The CI failure is likely due to https://github.com/scipy/scipy/pull/18631 which has just landed. The failure is in refguide-check, so I wonder why it did not fail in the PR. 
Anyhow, just `pip install threadpoolctl` or what else it needs.

**Comment by alugowski** [comment]
> The CI failure is likely due to #18631 which has just landed. The failure is in refguide-check, so I wonder why it did not fail in the PR. Anyhow, just `pip install threadpoolctl` or what else it needs.

Correct.

That PR adds `threadpoolctl` to the `pip install` command on `config.yml:80`, hence why it passes CI there. Pulling main should fix the issue here.

**Comment by ev-br** [comment]
Now that the CI is actually testing things (yay!), can work through failures. Quickly looking at several of them:

- Why is `array` event picked up here? It is not in the public API, so should not be even looked at with `strategy="api"`.

```
 [doctest] scipy.datasets._fetchers.array ___________________
EXAMPLE LOCATION UNKNOWN, not showing all tests of that example
??? >>> np.array([1, 2, 3])
UNEXPECTED EXCEPTION: NameError("name 'np' is not defined")
Traceback (most recent call last):
  File "/home/circleci/.pyenv/versions/3.9.18/lib/python3.9/doctest.py", line 1334, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest scipy.datasets._fetchers.array[0]>", line 1, in <module>
NameError: name 'np' is not defined
/home/circleci/.pyenv/versions/3.9.18/lib/python3.9/site-packages/numpy/__init__.py:None: UnexpectedException
```
I would compare the runs with pytest and manual, to see what's going wrong.


- This is legit: need to use raw strings, `r"Standard normal cumulative distribution function $\Phi$`, to shield backslashes:
```
  File "<doctest scipy.stats._qmvnt.phi[8]>", line 1
    ax.set_title("Standard normal cumulative distribution function $\Phi$")
                 ^
SyntaxError: invalid escape sequence \P
```

- This is also an error in the docstring, which now need an explicit `import numpy as np`:

```
142         Sample spacing (inverse of the sampling rate). Defaults to 1.
143 
144     Returns
145     -------
146     f : ndarray
147         Array of length `n` containing the sample frequencies.
148 
149     Examples
150     --------
151     >>> signal = np.array([-2, 8, 6, 4, 1, 0, 3, 5], dtype=float)
UNEXPECTED EXCEPTION: NameError("name 'np' is not defined")
```

- This one needs `parse_namedtuples=True` in the `DTConfig`, needs investigating why it is False:

```
E   NameError: name 'DescribeResult' is not defined
```

**Comment by ev-br** [comment]
3. is easiest: the function itself is deprecated, add to the filter

2. Maybe try copy-pasting the exact output? 

1. Is it really legal python? Try running the example in a python shell.

**Comment by ev-br** [comment]
Browsing through the CI log, I really like this one:

```
211 Construct a 1000x1000 `lil_array` and add some values to it:
212 
213 >>> from scipy.sparse import lil_array
214 >>> from scipy.sparse.linalg import spsolve
215 >>> from numpy.linalg import solve, norm
216 >>> from numpy.random import rand
217 
218 >>> A = lil_array((1000, 1000))
219 >>> A[0, :100] = rand(100)
220 >>> A[1, 100:200] = A[0, :100]
UNEXPECTED EXCEPTION: NotImplementedError('We have not yet implemented 1D sparse slices; please index using explicit indices, e.g. `x[:, [0]]`')
Traceback (most recent call last):
  File "/home/circleci/.pyenv/versions/3.9.18/lib/python3.9/doctest.py", line 1334, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest scipy.sparse[12]>", line 1, in <module>
  File "/home/circleci/repo/build-install/lib/python3.9/site-packages/scipy/sparse/_lil.py", line 150, in __getitem__
    return IndexMixin.__getitem__(self, key)
  File "/home/circleci/repo/build-install/lib/python3.9/site-packages/scipy/sparse/_index.py", line 53, in __getitem__
    self._raise_on_1d_array_slice()
  File "/home/circleci/repo/build-install/lib/python3.9/site-packages/scipy/sparse/_index.py", line 40, in _raise_on_1d_array_slice
    raise NotImplementedError(
NotImplementedError: We have not yet implemented 1D sparse slices; please index using explicit indices, e.g. `x[:, [0]]`
/home/circleci/repo/build-install/lib/python3.9/site-packages/scipy/sparse/__init__.py:220: UnexpectedException
```

So this looks like a genuine glitch in the documentation, which adverteses something not yet implemented?

**Comment by Sheila-nk** [comment]
I have been able to filter out DeprecationWarnings and somehow the Syntax errors from numpy ufuncs fixed themselves with the merging of this PR: https://github.com/ev-br/scpdt/pull/103
I will now focus my efforts on the output formatting errors.

**Comment by melissawm** [comment]
I still see the DeprecationWarnings locally with this version, should we pull in https://github.com/ev-br/scpdt/pull/103 to this PR? (I also edited the PR description to list the correct incantation for accuracy.

@Sheila-nk please let me know if you need help with anything here!

**Comment by Sheila-nk** [comment]
> I still see the DeprecationWarnings locally with this version, should we pull in https://github.com/ev-br/scpdt/pull/103 to this PR? (I also edited the PR description to list the correct incantation for accuracy.

Hmm, I wonder if it could be that you need to install the most recent version of scpdt since we have not listed the plugin as a scipy requirement so for now you might need to install it manually:
`pip install git+https://github.com/ev-br/scpdt.git@main` 

> please let me know if you need help with anything here!

Yes, I wonder if the current approach to manually change the tabs to spaces is fine, I remember you and @ev-br had mentioned something about a sphinx extension?

**Comment by ev-br** [comment]
For tabs->spaces I would very much recommend a low-tech solution. Just change them manually once and be done with it.

**Comment by lucascolley** [comment]
Looks like you need to merge and update the git submodules. This worked for me locally:
```
git checkout main
git submodule update
git checkout integrate-plugin
git merge main
cd scipy/_lib/unuran
git merge 1d315c6
cd ../../..
git add scipy/_lib/unuran
git commit
```

**Comment by Sheila-nk** [comment]
Thanks @lucascolley 😄 it worked!

**Comment by lucascolley** [comment]
hmm these CI errors are strange, it looks like your branch has some discrepancies with `main` in terms of e3bbb6b15752681532e48af769e52f401cd55c14 - I think something has gone wrong when resolving merge conflicts.

File: scipy/sparse/_extract.py
Comment (lucascolley) [inline scipy/sparse/_extract.py:96]:
```suggestion
    <3x5 sparse array of type '<class 'numpy.int32'>'
```

File: scipy/sparse/_matrix_io.py
Comment (lucascolley) [inline scipy/sparse/_matrix_io.py:114]: this should all be `array` rather than `matrix` - and whether or not we import scipy `as sp` or not needs to be consistent here and below.

**Review by lucascolley — CHANGES_REQUESTED** [review]
These should get `refguide-check` to pass.

edit: btw, you can use `[docs only]` now instead of `[skip actions] [skip cirrus]`, which will do the same thing except also run the linter

File: scipy/sparse/__init__.py
Comment (lucascolley) [inline scipy/sparse/__init__.py:233]: is this deliberate or has it slipped through the merge commit?

File: scipy/_lib/_testutils.py
Comment (lucascolley) [inline scipy/_lib/_testutils.py:81]: Looks like we're fine to restrict the `_lib` ignores to just the git submodules.

```suggestion
            pytest_args += [
                "--doctest-modules",
                "--ignore=scipy/interpolate/_interpnd_info.py",
                "--ignore=scipy/_lib/array_api_compat",
                "--ignore=scipy/_lib/highs",
                "--ignore=scipy/_lib/unuran",
            ]
```

**Review by lucascolley — COMMENTED** [review]
Looking very close, although I'm still getting 40 doctest errors (~30 with the extra `_lib` ignores)

**Comment by Sheila-nk** [comment]
Thank you @lucascolley for the review. Seems the refguide-check failures were due to the merge conflicts.

File: scipy/sparse/_extract.py
Comment (Sheila-nk) [reply scipy/sparse/_extract.py]: This was fixed after resolving the merge conflict. Thank you Lucas.

File: scipy/sparse/_extract.py
Comment (Sheila-nk) [reply scipy/sparse/_extract.py]: This was fixed after resolving the merge conflict. Thank you Lucas.

File: scipy/sparse/_matrix_io.py
Comment (Sheila-nk) [reply scipy/sparse/_matrix_io.py]: This was fixed after resolving the merge conflict. Thank you Lucas.

File: scipy/sparse/__init__.py
Comment (Sheila-nk) [reply scipy/sparse/__init__.py:233]: When running doctests through the plugin, this example was failing due to a `NotImplementedError` as mentioned here: https://github.com/scipy/scipy/pull/19242#issuecomment-1739410889
Explicit indexing currently solves this error. 
Is there a more effective method you might recommend to address the error?

File: scipy/io/matlab/_miobase.py
Comment (Sheila-nk) [reply scipy/io/matlab/_miobase.py:300]: When I run these examples in the Python shell, these are the results I get:
```python
>>> import numpy as np
>>> from scipy.io.matlab._miobase import matdims
>>> matdims(np.array([[]]))
(1, 0)
>>> matdims(np.array([[[]]]))
(1, 1, 0)
>>> 
```
You can confirm on your end.

Comment (lucascolley) [reply scipy/io/matlab/_miobase.py:300]: Yeah that's right, I just mean in case the behaviour of `matdims` is buggy and actually the previous example was correct (unlikely but possible)

**Comment by Sheila-nk** [comment]
@lucascolley maybe you can help with the NameErrors?
 I am unfamiliar with the functions and therefore do not know what the values of the variables could be.

1.
```python
___________ [doctest] scipy.stats._continuous_distns.reciprocal_gen ____________
8740 
8741     %(example)s
8742 
8743     This doesn't show the equal probability of ``0.01``, ``0.1`` and
8744     ``1``. This is best when the x-axis is log-scaled:
8745 
8746     >>> import numpy as np
8747     >>> import matplotlib.pyplot as plt
8748     >>> fig, ax = plt.subplots(1, 1)
8749     >>> ax.hist(np.log10(r))
UNEXPECTED EXCEPTION: NameError("name 'r' is not defined")
Traceback (most recent call last):
  File "/home/circleci/.pyenv/versions/3.11.7/lib/python3.11/doctest.py", line 1353, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest scipy.stats._continuous_distns.reciprocal_gen[3]>", line 1, in <module>
NameError: name 'r' is not defined
/home/circleci/repo/build-install/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:8749: UnexpectedException
```

2.
```python
_______________ [doctest] scipy.stats._discrete_distns.zipf_gen ________________
1284            https://en.wikipedia.org/wiki/Zeta_distribution
1285 
1286     %(example)s
1287 
1288     Confirm that `zipf` is the large `n` limit of `zipfian`.
1289 
1290     >>> import numpy as np
1291     >>> from scipy.stats import zipf, zipfian
1292     >>> k = np.arange(11)
1293     >>> np.allclose(zipf.pmf(k, a), zipfian.pmf(k, a, n=10000000))
UNEXPECTED EXCEPTION: NameError("name 'a' is not defined")
Traceback (most recent call last):
  File "/home/circleci/.pyenv/versions/3.11.7/lib/python3.11/doctest.py", line 1353, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest scipy.stats._discrete_distns.zipf_gen[3]>", line 1, in <module>
NameError: name 'a' is not defined
/home/circleci/repo/build-install/lib/python3.11/site-packages/scipy/stats/_discrete_distns.py:1293: UnexpectedException
```

File: scipy/_lib/_testutils.py
Comment (Sheila-nk) [reply scipy/_lib/_testutils.py:81]: Great! Okay, will make this change.

**Comment by lucascolley** [comment]
> maybe you can help with the NameErrors?
I am unfamiliar with the functions and therefore do not know what the values of the variables could be.

Hmm I could find values that work but not sure what the ideal values would be, probably best to ping @andyfaff as codeowner / @mdhaber

**Comment by lucascolley** [comment]
superseded by gh-20127 (thanks Sheila!)

**Pull request closed without merging**
